rpc/metadata.go

```diff
@@ -0,0 +1,42 @@
+package rpc
+
+import (
+	"context"
+	"net/http"
+)
+
+type rawMd struct {
+	headers http.Header
+}
+
+type mdOutgoingKey struct{}
+
+// NewOutgoingContext is used to attach http headers into the context
+func NewOutgoingContext(ctx context.Context, header http.Header) context.Context {
+	return context.WithValue(ctx, mdOutgoingKey{}, rawMd{headers: header})
+}
+
+// HeadersFromOutgoingContext is used to extract http headers from the context
+func HeadersFromOutgoingContext(ctx context.Context) (http.Header, bool) {
```

Comment:
This doesn't need to be public, does it?

Reply:
I don't see why the second return param is needed. The caller can just do a `nil`-check

Reply:
> This doesn't need to be public, does it?
I think it is necessary. users may have `append` requirements in different places.